vllm/model_executor/models/chatglm.py

```diff
@@ -0,0 +1,336 @@
+    return total_num_heads, num_heads
+
+
+def compute_tp_num_kv_heads(config, tp_world_size):
+    total_num_kv_heads = config.multi_query_group_num
+    if total_num_kv_heads >= tp_world_size:
+        # Number of KV heads is greater than TP size, so we partition
+        # the KV heads across multiple tensor parallel GPUs.
+        assert total_num_kv_heads % tp_world_size == 0
+    else:
+        # Number of KV heads is less than TP size, so we replicate
+        # the KV heads across multiple tensor parallel GPUs.
+        assert tp_world_size % total_num_kv_heads == 0
+    num_kv_heads = max(1, total_num_kv_heads // tp_world_size)
+    num_kv_heads_replicas = max(1, tp_world_size // total_num_kv_heads)
+    return total_num_kv_heads, num_kv_heads, num_kv_heads_replicas
+
+
+class Attention(nn.Module):
+
+    def __init__(self, config: ChatGLMConfig):
+        super().__init__()
+        self.rope_theta = 10000
+
+        tp_world_size = get_tensor_model_parallel_world_size()
+        self.total_num_heads, self.num_heads = compute_tp_num_heads(
+            config, tp_world_size)
+        self.head_dim = config.hidden_size // self.total_num_heads
+        self.total_num_kv_heads, self.num_kv_heads, num_kv_heads_replicas = \
+            compute_tp_num_kv_heads(config, tp_world_size)
+
+        self.attn = PagedAttentionWithRoPE(num_heads=self.num_heads,
+                                           head_size=self.head_dim,
+                                           scale=self.head_dim**-0.5,
+                                           rotary_dim=self.head_dim,
+                                           base=self.rope_theta,
+                                           is_glm_style=True,
+                                           is_neox_style=False,
+                                           num_kv_heads=self.num_kv_heads,
+                                           max_position=config.seq_length)
+
+        self.qkv_hidden_size = (
+            (self.total_num_heads +
+             2 * num_kv_heads_replicas * self.total_num_kv_heads) *
+            self.head_dim)
+        self.query_key_value = ColumnParallelLinear(
+            config.hidden_size,
+            self.qkv_hidden_size,
+            bias=config.add_bias_linear or config.add_qkv_bias,
+            gather_output=False,
+        )
+        self.dense = RowParallelLinear(
+            self.total_num_heads * self.head_dim,
+            config.hidden_size,
+            bias=config.add_bias_linear,
+            input_is_parallel=True,
+        )
+
+    def forward(
+        self,
+        positions: torch.Tensor,
+        hidden_states: torch.Tensor,
+        kv_cache: KVCache,
+        input_metadata: InputMetadata,
+        cache_event: Optional[torch.cuda.Event],
+    ) -> torch.Tensor:
+        hidden_states, _ = self.query_key_value(hidden_states)
+        query, key, value = hidden_states.split(
+            [
+                self.num_heads * self.head_dim,
```

Comment:
These variables may be precomputed in __init__() as `q_size, kv_size, kv_size` like llama impl.
`q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)`

Reply:
`**hidden_states**, _ = self.query_key_value(hidden_states)`

maybe `hidden_states` is not a good variable name to hold the return value